File: app/src/main/java/com/pr0gramm/app/ui/fragments/PostPagerFragment.java
package com.pr0gramm.app.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.common.base.Optional;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.R;
import com.pr0gramm.app.api.pr0gramm.response.Tag;
import com.pr0gramm.app.feed.Feed;
import com.pr0gramm.app.feed.FeedFilter;
import com.pr0gramm.app.feed.FeedItem;
import com.pr0gramm.app.feed.FeedLoader;
import com.pr0gramm.app.feed.FeedService;
import com.pr0gramm.app.feed.FeedType;
import com.pr0gramm.app.ui.FilterFragment;
import com.pr0gramm.app.ui.IdFragmentStatePagerAdapter;
import com.pr0gramm.app.ui.MainActionHandler;
import com.pr0gramm.app.ui.PreviewInfo;
import com.pr0gramm.app.ui.base.BaseFragment;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import javax.inject.Inject;

import butterknife.Bind;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.pr0gramm.app.ui.ScrollHideToolbarListener.ToolbarActivity;

/**
 */
public class PostPagerFragment extends BaseFragment implements FilterFragment {
    private static final Logger logger = LoggerFactory.getLogger("PostPagerFragment");

    private static final String ARG_FEED_PROXY = "PostPagerFragment.feedProxy";
    private static final String ARG_START_ITEM = "PostPagerFragment.startItem";
    private static final String ARG_START_ITEM_COMMENT = "PostPagerFragment.startItemComment";

    @Inject
    FeedService feedService;

    @Bind(R.id.pager)
    ViewPager viewPager;

    private Feed feed;
    private PostAdapter adapter;

    private PostFragment activePostFragment;

    private PreviewInfo previewInfo;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        return inflater.inflate(R.layout.fragment_post_pager, container, false);
    }

    @Override
    protected void injectComponent(ActivityComponent activityComponent) {
        activityComponent.inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // get the feed to show and setup a loader to load more data
        feed = getArgumentFeed(savedInstanceState);
        FeedLoader loader = new FeedLoader(FeedLoader.bindTo(this, this::onLoadError), feedService, feed);

        // create the adapter on the view
        adapter = new PostAdapter(getChildFragmentManager(), feed, loader) {
            @Override
            public void setPrimaryItem(ViewGroup container, int position, Object object) {
                super.setPrimaryItem(container, position, object);
                updateActiveItem((PostFragment) object);
                saveStateToBundle(getArguments());
            }
        };

        if (getActivity() instanceof ToolbarActivity) {
            ToolbarActivity activity = (ToolbarActivity) getActivity();
            activity.getScrollHideToolbarListener().reset();

            viewPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
                @Override
                public void onPageScrollStateChanged(int state) {
                    activity.getScrollHideToolbarListener().reset();
                    if (activePostFragment != null) {
                        activePostFragment.exitFullscreen();
                    }
                }
            });
        }

        viewPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                if (position >= 0 && position + 1 < adapter.getCount()) {
                    Optional<Fragment> prev = adapter.getFragment(position);
                    Optional<Fragment> next = adapter.getFragment(position + 1);
                    if (prev.isPresent() && next.isPresent()) {
                        ((PostFragment) prev.get()).mediaHorizontalOffset(positionOffsetPixels / 2);
                        ((PostFragment) next.get()).mediaHorizontalOffset(
                                -viewPager.getWidth() / 2 + positionOffsetPixels / 2);
                    }
                }
            }
        });

        viewPager.setAdapter(adapter);

        if (savedInstanceState != null) {
            // calculate index of the first item to show if this is the first
            // time we show this fragment.
            FeedItem start = getArgumentStartItem(savedInstanceState);
            makeItemCurrent(start);
        }
    }

    private void onLoadError(Throwable throwable) {
        logger.warn("Could not load feed.", throwable);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);

        // calculate index of the first item to show if this is the first
        // time we show this fragment.
        FeedItem start = getArgumentStartItem(savedInstanceState);
        makeItemCurrent(start);
    }

    private void makeItemCurrent(FeedItem item) {
        int index = feed.indexOf(item).or(0);

        logger.info("Moving to index: " + index);
        viewPager.setCurrentItem(index, false);
    }

    private void updateActiveItem(PostFragment newActiveFragment) {
        int position = adapter.getItemPosition(newActiveFragment);
        if (activePostFragment == newActiveFragment)
            return;

        logger.info("Setting feed item activate at " + position + " to " + newActiveFragment);

        // deactivate previous item
        if (activePostFragment != null)
            activePostFragment.setActive(false);

        // and activate the next one
        activePostFragment = newActiveFragment;
        if (activePostFragment != null) {
            activePostFragment.setActive(true);

            // try scroll to initial comment. This will only work if the comment
            // is a part of the given post and will otherwise do nothing
            long startCommentId = getArguments().getLong(ARG_START_ITEM_COMMENT);
            if (startCommentId > 0) {
                activePostFragment.autoScrollToComment(startCommentId);
            }

            if (previewInfo != null) {
                if (activePostFragment.getFeedItem().getId() == getArgumentStartItem(null).getId()) {
                    activePostFragment.setPreviewInfo(previewInfo);
                    previewInfo = null;
                }
            }
        }
    }

    /**
     * Gets the feed from the saved state. If there is no state
     * or it does not contain the feed proxy, the feed proxy is extracted
     * from {@link #getArguments()}
     *
     * @param savedState An optional saved state.
     */
    private Feed getArgumentFeed(@Nullable Bundle savedState) {
        Bundle encoded = null;
        if (savedState != null)
            encoded = savedState.getBundle(ARG_FEED_PROXY);

        if (encoded == null)
            encoded = getArguments().getBundle(ARG_FEED_PROXY);

        return Feed.restore(checkNotNull(encoded, "No feed-proxy found"));
    }

    /**
     * @see #getArgumentFeed(Bundle)
     */
    private FeedItem getArgumentStartItem(@Nullable Bundle saveState) {
        FeedItem result = null;

        if (saveState != null)
            result = saveState.getParcelable(ARG_START_ITEM);

        if (result == null)
            result = getArguments().getParcelable(ARG_START_ITEM);

        return checkNotNull(result, "No start-item found");
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        saveStateToBundle(outState);
    }

    /**
     * Returns the feed filter for this fragment.
     */
    @Override
    public FeedFilter getCurrentFilter() {
        // prevent errors here
        if (feed == null)
            return new FeedFilter();

        return feed.getFeedFilter();
    }

    public void onTagClicked(Tag tag) {
        ((MainActionHandler) getActivity()).onFeedFilterSelected(
                getCurrentFilter().withTags(tag.getTag()));
    }

    public void onUsernameClicked(String username) {
        // always show all uploads of a user.
        FeedFilter newFilter = getCurrentFilter()
                .withFeedType(FeedType.NEW)
                .withUser(username);

        ((MainActionHandler) getActivity()).onFeedFilterSelected(newFilter);
    }

    private void saveStateToBundle(Bundle outState) {
        if (viewPager != null && feed != null) {
            int position = viewPager.getCurrentItem();
            FeedItem item = feed.at(position);
            outState.putParcelable(ARG_START_ITEM, item);
            outState.putParcelable(ARG_FEED_PROXY, feed.persist(position));
        }
    }

    /**
     * Sets the pixels that should be used in the transition.
     */
    public void setPreviewInfo(PreviewInfo previewInfo) {
        this.previewInfo = previewInfo;
    }

    private static class PostAdapter extends IdFragmentStatePagerAdapter implements Feed.FeedListener {
        private final Feed proxy;
        private final FeedLoader loader;

        public PostAdapter(FragmentManager fragmentManager, Feed proxy, FeedLoader loader) {
            super(fragmentManager);

            this.proxy = proxy;
            this.proxy.setFeedListener(this);
            this.loader = loader;
        }

        @Override
        public Fragment getItem(int position) {
            if (!loader.isLoading()) {
                if (position > proxy.size() - 12) {
                    logger.info("requested pos=" + position + ", load next page");
                    loader.next();
                }

                if (position < 12) {
                    logger.info("requested pos=" + position + ", load prev page");
                    loader.previous();
                }
            }

            return PostFragment.newInstance(proxy.at(position));
        }

        @Override
        public int getCount() {
            return proxy.size();
        }

        @Override
        public int getItemPosition(Object object) {
            FeedItem item = ((PostFragment) object).getFeedItem();
            return proxy.indexOf(item).or(PagerAdapter.POSITION_NONE);
        }

        @Override
        protected long getItemId(int position) {
            return proxy.at(position).getId();
        }

        @Override
        public void onNewItems(List<FeedItem> newItems) {
            notifyDataSetChanged();
        }

        @Override
        public void onRemoveItems() {
            // should not happen
            throw new UnsupportedOperationException();
        }
    }

    public static PostPagerFragment newInstance(Feed feed, int idx, Optional<Long> commentId) {
        Bundle arguments = new Bundle();
        arguments.putBundle(ARG_FEED_PROXY, feed.persist(idx));
        arguments.putParcelable(ARG_START_ITEM, feed.at(idx));
        arguments.putLong(ARG_START_ITEM_COMMENT, commentId.or(-1L));

        PostPagerFragment fragment = new PostPagerFragment();
        fragment.setArguments(arguments);
        return fragment;
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/fragments/PostPagerFragment.kt
package com.pr0gramm.app.ui.fragments

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.commitNow
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.viewpager.widget.ViewPager
import com.pr0gramm.app.R
import com.pr0gramm.app.Settings
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.databinding.FragmentPostPagerBinding
import com.pr0gramm.app.feed.*
import com.pr0gramm.app.parcel.getParcelableOrNull
import com.pr0gramm.app.ui.*
import com.pr0gramm.app.ui.ScrollHideToolbarListener.ToolbarActivity
import com.pr0gramm.app.ui.base.BaseFragment
import com.pr0gramm.app.ui.base.bindViews
import com.pr0gramm.app.ui.base.launchUntilViewDestroy
import com.pr0gramm.app.ui.fragments.feed.FeedFragment
import com.pr0gramm.app.ui.fragments.post.PostFragment
import com.pr0gramm.app.util.arguments
import com.pr0gramm.app.util.di.instance
import com.pr0gramm.app.util.observeChangeEx
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.filterIsInstance

/**
 */
class PostPagerFragment : BaseFragment("PostPagerFragment", R.layout.fragment_post_pager), FilterFragment, TitleFragment, PreviewInfoSource {
    private val feedService: FeedService by instance()

    private val views by bindViews(FragmentPostPagerBinding::bind)

    private lateinit var adapter: PostAdapter

    private var previewInfo: PreviewInfo? = null

    private var latestActivePostFragment: PostFragment? = null

    // to prevent double saving of current state
    private var lastSavedPosition: Int = -1

    private var initialCommentRef: CommentRef? = null

    private val settings = Settings.get()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initialCommentRef = arguments?.getParcelable(ARG_START_ITEM_COMMENT_REF)

        // Listen for changes in fragments. We use this, because with the ViewPager2
        // callbacks we don't have any access to the fragment, and we don't even know,
        // when the fragment will be there and available.
        val l = object : FragmentManager.FragmentLifecycleCallbacks() {
            override fun onFragmentResumed(fm: FragmentManager, f: Fragment) {
                logger.debug { "Fragment was resumed." }
                if (f is PostFragment && latestActivePostFragment !== f) {
                    latestActivePostFragment = f
                    onActiveFragmentChanged()
                }
            }
        }

        childFragmentManager.registerFragmentLifecycleCallbacks(l, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // get the feed to show and setup a loader to load more data
        val previousFeed = getArgumentFeed(savedInstanceState)
        val manager = FeedManager(viewLifecycleOwner.lifecycleScope, feedService, previousFeed)

        // create the adapter on the view
        adapter = PostAdapter(manager, previousFeed)

        launchUntilViewDestroy {
            manager.updates.filterIsInstance<FeedManager.Update.NewFeed>().collect { update ->
                adapter.feed = update.feed
            }
        }

        if (activity is ToolbarActivity) {
            val activity = activity as ToolbarActivity
            activity.scrollHideToolbarListener.reset()

            views.pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrollStateChanged(state: Int) {
                    activity.scrollHideToolbarListener.reset()
                    latestActivePostFragment?.exitFullscreen()
                }

                override fun onPageSelected(position: Int) {
                    logger.info { "Page was selected: $position" }

                    val fragment = adapter.getFragment(position)
                    if (fragment != null && fragment != latestActivePostFragment) {
                        childFragmentManager.commitNow(allowStateLoss = true) {
                            val previousFragment = latestActivePostFragment
                            if (previousFragment != null && previousFragment.isAdded) {
                                setMaxLifecycle(previousFragment, Lifecycle.State.STARTED)
                            }

                            setMaxLifecycle(fragment, Lifecycle.State.RESUMED)
                        }
                    }
                }

                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
            })
        }

        if (Settings.get().fancyScrollHorizontal) {
            views.pager.setPageTransformer(false) { page: View, position: Float ->
                val viewer = page.findViewWithTag<View>(PostFragment.ViewerTag)
                if (viewer != null) {
                    viewer.translationX = -(position * page.width / 2.0f)
                }
            }
        }

        views.pager.adapter = adapter
        views.pager.offscreenPageLimit = 1

        // calculate index of the first item to show if this is the first
        // time we show this fragment.
        makeItemCurrent(getArgumentStartItem(savedInstanceState))
    }

    private fun onActiveFragmentChanged() {
        val mainActivity = activity as? MainActivity
        mainActivity?.updateActionbarTitle()
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)

        val item = getArgumentStartItem(savedInstanceState)

        if (adapter.feed.getOrNull(views.pager.currentItem)?.id != item.id) {
            // the position of the view pager might have been saved. We need to re-save it now.
            logger.info { "Going to restore view position in onViewStateRestored" }
            makeItemCurrent(item)
        }
    }

    override fun onStop() {
        latestActivePostFragment?.feedItem?.let { feedItem ->
            val feed = adapter.feed
            if (feedItem in feed) {
                val target = targetFragment as? FeedFragment
                        ?: return@let
                target.updateFeedItemTarget(adapter.feed, feedItem)
            }
        }

        super.onStop()
    }

    private fun makeItemCurrent(item: FeedItem) {
        val index = adapter.feed.indexById(item.id) ?: 0

        logger.info { "Restore feed at index: $index (${item.id})" }
        views.pager.setCurrentItem(index, false)
    }

    override fun previewInfoFor(item: FeedItem): PreviewInfo? {
        return previewInfo?.takeIf { it.itemId == item.id }
    }

    /**
     * Get the feed from the given bundle.
     */
    private fun getArgumentFeed(savedState: Bundle?): Feed {
        val parceled = savedState?.getParcelableOrNull<Feed.FeedParcel>(ARG_FEED)
                ?: arguments?.getParcelableOrNull<Feed.FeedParcel>(ARG_FEED)
                ?: throw IllegalStateException("No feed found.")

        return parceled.feed
    }

    /**
     * @see getArgumentFeed
     */
    private fun getArgumentStartItem(savedState: Bundle?): FeedItem {
        return savedState?.getParcelableOrNull<FeedItem>(ARG_START_ITEM)
                ?: arguments?.getParcelableOrNull<FeedItem>(ARG_START_ITEM)
                ?: throw IllegalStateException("No initial item found.")
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        saveStateToBundle(outState)
    }

    /**
     * Returns the feed filter for this fragment.
     */
    override val currentFilter: FeedFilter
        get() = adapter.feed.filter

    override val title: TitleFragment.Title?
        get() = buildFragmentTitle()

    private fun buildFragmentTitle(): TitleFragment.Title? {
        val titleOverride = arguments?.getString(ARG_TITLE)

        if (settings.useTopTagAsTitle) {
            // fetch title from the current active post fragment
            var title = latestActivePostFragment?.title

            if (titleOverride != null) {
                // if the caller gave us a more specific title, we'll use that one.
                title = title?.copy(title = titleOverride) ?: TitleFragment.Title(titleOverride)
            }

            return title
        } else {
            // build one from the current filter
            val context = this.context ?: return null
            return FeedFilterFormatter.toTitle(context, currentFilter)
        }
    }

    fun onTagClicked(tag: Api.Tag) {
        val handler = activity as MainActionHandler
        handler.onFeedFilterSelected(currentFilter.withTags(tag.text))
    }

    fun onUsernameClicked(username: String) {
        // always show all uploads of a user.
        val newFilter = currentFilter
                .withFeedType(FeedType.NEW)
                .withUser(username)

        (activity as MainActionHandler).onFeedFilterSelected(newFilter)
    }

    private fun saveStateToBundle(outState: Bundle) {
        if (view != null && adapter.feed.isNotEmpty()) {
            val position = views.pager.currentItem.coerceIn(adapter.feed.indices)

            if (lastSavedPosition != position) {
                lastSavedPosition = position

                val item = adapter.feed[position]
                outState.putParcelable(ARG_START_ITEM, item)
                outState.putParcelable(ARG_FEED, adapter.feed.parcelAround(position))

                logger.debug { "Saved $position (id=${item.id})" }
            }
        }
    }

    /**
     * Sets the pixels that should be used in the transition.
     */
    fun setPreviewInfo(previewInfo: PreviewInfo) {
        this.previewInfo = previewInfo
    }

    private inner class PostAdapter(
            private val manager: FeedManager,
            feed: Feed)
        : IdFragmentStatePagerAdapter(childFragmentManager) {

        var feed: Feed by observeChangeEx(feed) { oldValue, newValue ->
            notifyDataSetChanged()
        }

        override fun setPrimaryItem(container: ViewGroup, position: Int, `object`: Any) {
            super.setPrimaryItem(container, position, `object`)
            arguments?.let { args -> saveStateToBundle(args) }
        }

        override fun getItem(position: Int): Fragment {
            if (!manager.isLoading) {
                if (position > feed.size - 12) {
                    logger.debug { "Requested pos=$position, load next page" }
                    manager.next()
                }

                if (position < 12) {
                    logger.debug { "Requested pos=$position, load prev page" }
                    manager.previous()
                }
            }

            // build a new fragment from the given item.
            val capped = position.coerceIn(0, feed.size - 1)
            val item = feed[capped]

            // initialize with reference to comment
            val initialCommentRef = initialCommentRef
            if (initialCommentRef?.itemId == item.id) {
                this@PostPagerFragment.initialCommentRef = null
                return PostFragment.newInstance(item, initialCommentRef)
            }

            return PostFragment.newInstance(item)
        }

        override fun getCount(): Int {
            return feed.size
        }

        override fun getItemPosition(`object`: Any): Int {
            val item = (`object` as PostFragment).feedItem
            return feed.indexById(item.id) ?: POSITION_NONE
        }

        override fun getItemId(position: Int): Long {
            return feed[position].id
        }
    }

    companion object {
        private const val ARG_FEED = "PP.feed"
        private const val ARG_TITLE = "PP.title"
        private const val ARG_START_ITEM = "PP.startItem"
        private const val ARG_START_ITEM_COMMENT_REF = "PP.startItemComment"

        fun newInstance(feed: Feed, idx: Int, commentRef: CommentRef?, fragmentTitle: String?): PostPagerFragment {
            return PostPagerFragment().arguments {
                putParcelable(ARG_FEED, feed.parcelAround(idx))
                putParcelable(ARG_START_ITEM, feed[idx])
                putParcelable(ARG_START_ITEM_COMMENT_REF, commentRef)
                putString(ARG_TITLE, fragmentTitle)
            }
        }
    }
}
